File: bot.js
const { Client, GatewayIntentBits, SlashCommandBuilder, REST, Routes } = require('discord.js');
const {
  joinVoiceChannel,
  createAudioPlayer,
  createAudioResource,
  AudioPlayerStatus,
  VoiceConnectionStatus,
  getVoiceConnection,
  entersState,
} = require('@discordjs/voice');
const gtts = require('gtts');
const fs = require('fs');
const path = require('path');
const { v4: uuidv4 } = require('uuid');

require('dotenv').config();

const client = new Client({
  intents: [
    GatewayIntentBits.Guilds,
    GatewayIntentBits.GuildMessages,
    GatewayIntentBits.MessageContent,
    GatewayIntentBits.GuildVoiceStates,
    GatewayIntentBits.GuildMembers,
  ],
});

// State management
const guildState = new Map(); // guildId -> { connection, player, queue, autoJoin, textChannelId }

function getState(guildId) {
  if (!guildState.has(guildId)) {
    guildState.set(guildId, {
      connection: null,
      player: createAudioPlayer(),
      queue: [],
      autoJoin: false,
      textChannelId: null,
      processing: false,
    });

    const state = guildState.get(guildId);

    state.player.on(AudioPlayerStatus.Idle, () => {
      processQueue(guildId);
    });

    state.player.on('error', (err) => {
      console.error(`[Player Error] Guild ${guildId}:`, err.message);
      processQueue(guildId);
    });
  }
  return guildState.get(guildId);
}

async function speakText(text, guildId) {
  return new Promise((resolve, reject) => {
    const tmpFile = path.join('/tmp', `tts_${uuidv4()}.mp3`);
    const tts = new gtts(text, 'en');
    tts.save(tmpFile, async (err) => {
      if (err) return reject(err);

      const state = getState(guildId);
      if (!state.connection) {
        fs.unlink(tmpFile, () => {});
        return resolve();
      }

      const resource = createAudioResource(tmpFile);
      state.queue.push({ resource, file: tmpFile });

      if (!state.processing) {
        processQueue(guildId);
      }
      resolve();
    });
  });
}

function processQueue(guildId) {
  const state = getState(guildId);
  if (state.queue.length === 0) {
    state.processing = false;
    return;
  }

  state.processing = true;
  const { resource, file } = state.queue.shift();

  if (!state.connection) {
    state.processing = false;
    state.queue = [];
    return;
  }

  try {
    state.connection.subscribe(state.player);
    state.player.play(resource);

    // Clean up file after a delay
    setTimeout(() => {
      fs.unlink(file, () => {});
    }, 10000);
  } catch (err) {
    console.error('[Queue Error]', err.message);
    state.processing = false;
  }
}

async function joinChannel(voiceChannel, guildId, textChannelId) {
  try {
    const connection = joinVoiceChannel({
      channelId: voiceChannel.id,
      guildId: guildId,
      adapterCreator: voiceChannel.guild.voiceAdapterCreator,
      selfDeaf: false,
    });

    await entersState(connection, VoiceConnectionStatus.Ready, 10_000);

    const state = getState(guildId);
    state.connection = connection;
    state.textChannelId = textChannelId;

    connection.on(VoiceConnectionStatus.Disconnected, async () => {
      try {
        await Promise.race([
          entersState(connection, VoiceConnectionStatus.Signalling, 5_000),
          entersState(connection, VoiceConnectionStatus.Connecting, 5_000),
        ]);
      } catch {
        connection.destroy();
        state.connection = null;
        state.queue = [];
        state.processing = false;
      }
    });

    return connection;
  } catch (err) {
    console.error('[Join Error]', err.message);
    throw err;
  }
}

// Register slash commands
async function registerCommands() {
  const commands = [
    new SlashCommandBuilder()
      .setName('join')
      .setDescription('Bot joins your current voice channel'),
    new SlashCommandBuilder()
      .setName('leave')
      .setDescription('Bot leaves the voice channel'),
    new SlashCommandBuilder()
      .setName('autojoin')
      .setDescription('Toggle auto-join when someone types in this channel'),
  ].map((cmd) => cmd.toJSON());

  const rest = new REST({ version: '10' }).setToken(process.env.DISCORD_TOKEN);

  try {
    console.log('Registering slash commands...');
    await rest.put(Routes.applicationCommands(process.env.CLIENT_ID), { body: commands });
    console.log('✅ Slash commands registered globally.');
  } catch (err) {
    console.error('Failed to register commands:', err);
  }
}

client.once('ready', async () => {
  console.log(`✅ Logged in as ${client.user.tag}`);
  await registerCommands();
});

// Handle slash commands
client.on('interactionCreate', async (interaction) => {
  if (!interaction.isChatInputCommand()) return;

  const { commandName, guild, member } = interaction;
  const guildId = guild.id;
  const state = getState(guildId);

  if (commandName === 'join') {
    const voiceChannel = member.voice?.channel;
    if (!voiceChannel) {
      return interaction.reply({ content: '❌ You need to be in a voice channel first!', ephemeral: true });
    }

    try {
      await joinChannel(voiceChannel, guildId, interaction.channelId);
      await interaction.reply(`✅ Joined **${voiceChannel.name}**! I'll speak everything typed in this channel.`);
    } catch {
      await interaction.reply({ content: '❌ Failed to join the voice channel.', ephemeral: true });
    }
  }

  else if (commandName === 'leave') {
    const connection = getVoiceConnection(guildId);
    if (connection) {
      connection.destroy();
      state.connection = null;
      state.queue = [];
      state.processing = false;
      state.autoJoin = false;
      await interaction.reply('👋 Left the voice channel.');
    } else {
      await interaction.reply({ content: "❌ I'm not in a voice channel.", ephemeral: true });
    }
  }

  else if (commandName === 'autojoin') {
    state.autoJoin = !state.autoJoin;
    state.textChannelId = interaction.channelId;
    const status = state.autoJoin ? '✅ **Auto-join enabled**' : '❌ **Auto-join disabled**';
    await interaction.reply(`${status} — I'll ${state.autoJoin ? 'automatically join your voice channel' : 'no longer auto-join'} when you type here.`);
  }
});

// Handle messages — speak ALL messages (no command needed)
client.on('messageCreate', async (message) => {
  if (message.author.bot) return;
  if (!message.guild) return;

  const guildId = message.guild.id;
  const state = getState(guildId);

  // Auto-join logic: if autojoin is on and someone types in the linked channel
  if (state.autoJoin && message.channelId === state.textChannelId && !state.connection) {
    const voiceChannel = message.member?.voice?.channel;
    if (voiceChannel) {
      try {
        await joinChannel(voiceChannel, guildId, message.channelId);
        console.log(`[AutoJoin] Joined ${voiceChannel.name} in ${message.guild.name}`);
      } catch {
        return;
      }
    }
  }

  // Only speak if bot is in a voice channel in this guild
  if (!state.connection) return;

  // Only speak messages from the linked text channel
  if (state.textChannelId && message.channelId !== state.textChannelId) return;

  // Skip messages that start with ! or / (commands)
  const content = message.content.trim();
  if (!content || content.startsWith('/') || content.startsWith('!')) return;

  // Clean up message for TTS
  let text = content
    .replace(/<@!?\d+>/g, '') // remove mentions
    .replace(/<#\d+>/g, '')   // remove channel mentions
    .replace(/<:\w+:\d+>/g, '') // remove custom emojis
    .replace(/https?:\/\/\S+/g, 'link') // replace URLs with "link"
    .replace(/\n+/g, '. ')
    .trim();

  if (!text) return;

  // Prepend username
  const displayName = message.member?.displayName || message.author.username;
  const ttsText = `${displayName} says: ${text}`;

  try {
    await speakText(ttsText, guildId);
  } catch (err) {
    console.error('[TTS Error]', err.message);
  }
});

// Handle voice state updates (member joins VC while autojoin is on)
client.on('voiceStateUpdate', async (oldState, newState) => {
  if (newState.member?.user.bot) return;

  const guildId = newState.guild.id;
  const state = getState(guildId);

  // If autojoin and user joins a VC and bot isn't in one
  if (state.autoJoin && newState.channelId && !oldState.channelId && !state.connection) {
    try {
      await joinChannel(newState.channel, guildId, state.textChannelId);
      console.log(`[AutoJoin] Joined ${newState.channel.name} on VC join`);
    } catch {
      // Silently fail
    }
  }
});

client.login(process.env.DISCORD_TOKEN);

File: bot.py
import discord
from discord.ext import commands
from discord import app_commands
import asyncio
import os
import tempfile
from gtts import gTTS

# ── Config ────────────────────────────────────────────────────────────────────
TOKEN  = os.environ["DISCORD_TOKEN"]
PREFIX = os.getenv("PREFIX", "!")
# ─────────────────────────────────────────────────────────────────────────────

intents = discord.Intents.default()
intents.message_content = True

bot = commands.Bot(command_prefix=PREFIX, intents=intents)

LANGUAGES = {
    "english":    "en",
    "hindi":      "hi",
    "french":     "fr",
    "spanish":    "es",
    "german":     "de",
    "japanese":   "ja",
    "korean":     "ko",
    "arabic":     "ar",
    "russian":    "ru",
    "italian":    "it",
    "portuguese": "pt",
    "chinese":    "zh",
}

# guild_id -> { "channel_id": int, "lang": str }
tts_channels: dict[int, dict] = {}


# ── Helpers ───────────────────────────────────────────────────────────────────

async def generate_tts(text: str, lang: str = "en") -> str:
    loop = asyncio.get_event_loop()
    def _gen():
        tts = gTTS(text=text, lang=lang)
        tmp = tempfile.NamedTemporaryFile(delete=False, suffix=".mp3", dir="/tmp")
        tts.save(tmp.name)
        return tmp.name
    return await loop.run_in_executor(None, _gen)


async def get_voice(guild: discord.Guild, member: discord.Member = None) -> discord.VoiceClient | None:
    vc = guild.voice_client
    if member and member.voice:
        channel = member.voice.channel
        if vc and vc.channel != channel:
            await vc.move_to(channel)
        elif not vc:
            vc = await channel.connect()
    return guild.voice_client


def play_file(vc: discord.VoiceClient, filepath: str):
    # Queue if already playing
    if vc.is_playing():
        vc.stop()

    def after(err):
        try:
            os.remove(filepath)
        except Exception:
            pass

    vc.play(discord.FFmpegPCMAudio(filepath), after=after)


# ── Events ────────────────────────────────────────────────────────────────────

@bot.event
async def on_ready():
    print(f"✅  {bot.user} is online  |  Servers: {len(bot.guilds)}")
    await bot.change_presence(activity=discord.Activity(
        type=discord.ActivityType.listening,
        name="your messages 👂"
    ))
    try:
        synced = await bot.tree.sync()
        print(f"   Synced {len(synced)} slash command(s)")
    except Exception as e:
        print(f"   Slash sync error: {e}")


@bot.event
async def on_message(message: discord.Message):
    # Ignore bots and DMs
    if message.author.bot or not message.guild:
        await bot.process_commands(message)
        return

    cfg = tts_channels.get(message.guild.id)

    # If auto-TTS is active and message is in the TTS channel
    if cfg and message.channel.id == cfg["channel_id"]:
        # Skip command messages
        if message.content.startswith(PREFIX):
            await bot.process_commands(message)
            return

        text = message.content.strip()
        if not text:
            await bot.process_commands(message)
            return

        vc = message.guild.voice_client
        if not vc:
            # Try to join the author's voice channel
            if message.author.voice:
                vc = await message.author.voice.channel.connect()
            else:
                await message.add_reaction("❌")
                await bot.process_commands(message)
                return

        lang = cfg.get("lang", "en")
        filepath = await generate_tts(text, lang)
        play_file(vc, filepath)
        await message.add_reaction("🔊")

    await bot.process_commands(message)


# ── Prefix Commands ───────────────────────────────────────────────────────────

@bot.command(name="join", help="Join your voice channel.")
async def join(ctx: commands.Context):
    if not ctx.author.voice:
        return await ctx.send("❌ Join a voice channel first.")
    vc = await get_voice(ctx.guild, ctx.author)
    await ctx.send(f"🔊 Joined **{vc.channel.name}**")


@bot.command(name="leave", aliases=["dc"], help="Leave the voice channel and disable auto-TTS.")
async def leave(ctx: commands.Context):
    tts_channels.pop(ctx.guild.id, None)
    if ctx.voice_client:
        await ctx.voice_client.disconnect()
        await ctx.send("👋 Disconnected and auto-TTS disabled.")
    else:
        await ctx.send("❌ Not in a voice channel.")


@bot.command(name="setup", help="Set this channel as the auto-TTS channel. Usage: !setup [language]")
async def setup(ctx: commands.Context, language: str = "english"):
    if not ctx.author.voice:
        return await ctx.send("❌ Join a voice channel first.")

    lang = LANGUAGES.get(language.lower(), "en")
    vc = await get_voice(ctx.guild, ctx.author)

    tts_channels[ctx.guild.id] = {
        "channel_id": ctx.channel.id,
        "lang": lang
    }

    await ctx.send(
        f"✅ Auto-TTS enabled!\n"
        f"🔊 Voice: **{vc.channel.name}**\n"
        f"💬 Text channel: **#{ctx.channel.name}**\n"
        f"🌐 Language: **{language}**\n\n"
        f"Just type anything here and I'll say it!"
    )


@bot.command(name="setlang", help="Change TTS language. Usage: !setlang hindi")
async def setlang(ctx: commands.Context, language: str = "english"):
    if ctx.guild.id not in tts_channels:
        return await ctx.send("❌ Run `!setup` first.")
    lang = LANGUAGES.get(language.lower())
    if not lang:
        return await ctx.send(f"❌ Unknown language. Use `!langs` to see options.")
    tts_channels[ctx.guild.id]["lang"] = lang
    await ctx.send(f"✅ Language set to **{language}**.")


@bot.command(name="stop", help="Stop current TTS playback.")
async def stop(ctx: commands.Context):
    if ctx.voice_client and ctx.voice_client.is_playing():
        ctx.voice_client.stop()
        await ctx.send("⏹️ Stopped.")
    else:
        await ctx.send("❌ Nothing is playing.")


@bot.command(name="langs", help="Show supported languages.")
async def langs(ctx: commands.Context):
    desc = "\n".join(f"`{k}` → `{v}`" for k, v in LANGUAGES.items())
    embed = discord.Embed(title="🌐 Supported Languages", description=desc, color=0x5865F2)
    await ctx.send(embed=embed)


@bot.command(name="ping", help="Check bot latency.")
async def ping(ctx: commands.Context):
    await ctx.send(f"🏓 Pong! `{round(bot.latency * 1000)}ms`")


# ── Slash Commands ────────────────────────────────────────────────────────────

@bot.tree.command(name="setup", description="Enable auto-TTS: bot reads everything typed in this channel.")
@app_commands.describe(language="Language to speak in (default: english)")
async def slash_setup(interaction: discord.Interaction, language: str = "english"):
    await interaction.response.defer()
    if not interaction.user.voice:
        return await interaction.followup.send("❌ Join a voice channel first.")

    lang = LANGUAGES.get(language.lower(), "en")
    vc = await get_voice(interaction.guild, interaction.user)

    tts_channels[interaction.guild.id] = {
        "channel_id": interaction.channel.id,
        "lang": lang
    }

    await interaction.followup.send(
        f"✅ Auto-TTS enabled!\n"
        f"🔊 Voice: **{vc.channel.name}**\n"
        f"💬 Text channel: **#{interaction.channel.name}**\n"
        f"🌐 Language: **{language}**\n\n"
        f"Just type anything here and I'll say it!"
    )


@bot.tree.command(name="leave", description="Disconnect and disable auto-TTS.")
async def slash_leave(interaction: discord.Interaction):
    tts_channels.pop(interaction.guild.id, None)
    vc = interaction.guild.voice_client
    if vc:
        await vc.disconnect()
        await interaction.response.send_message("👋 Disconnected and auto-TTS disabled.")
    else:
        await interaction.response.send_message("❌ Not in a voice channel.")


@bot.tree.command(name="stop", description="Stop current TTS playback.")
async def slash_stop(interaction: discord.Interaction):
    vc = interaction.guild.voice_client
    if vc and vc.is_playing():
        vc.stop()
        await interaction.response.send_message("⏹️ Stopped.")
    else:
        await interaction.response.send_message("❌ Nothing is playing.")


# ── Run ───────────────────────────────────────────────────────────────────────

bot.run(TOKEN)
